src/save-restore.c

```diff
@@ -0,0 +1,438 @@
+save_hint_cb (XdpDbusImplSaveRestore *impl,
+              const char             *session_id,
+              gpointer                data)
+{
+  XdpSaveRestore *self = XDP_SAVE_RESTORE (data);
+  g_autoptr(XdpSession) session = xdp_session_lookup (session_id);
+  XdpSaveRestoreSession *save_restore_session = XDP_SAVE_RESTORE_SESSION_SESSION (session);
+
+  if (save_restore_session && !save_restore_session->closed)
+    xdp_dbus_save_restore_emit_save_hint (XDP_DBUS_SAVE_RESTORE (self),
+                                          session->id);
+}
+
+static void
+quit_cb (XdpDbusImplSaveRestore *impl,
+              const char             *session_id,
+              gpointer                data)
+{
+  XdpSaveRestore *self = XDP_SAVE_RESTORE (data);
+  g_autoptr(XdpSession) session = xdp_session_lookup (session_id);
+  XdpSaveRestoreSession *save_restore_session = XDP_SAVE_RESTORE_SESSION_SESSION (session);
+
+  if (save_restore_session && !save_restore_session->closed)
+    xdp_dbus_save_restore_emit_save_hint (XDP_DBUS_SAVE_RESTORE (self),
+                                          session->id);
+}
+
+static void
+xdp_save_restore_set_impl (XdpSaveRestore         *self,
+                           XdpDbusImplSaveRestore *new_impl)
+{
+  if (self->impl == new_impl)
+    return;
+
+  if (self->impl)
+    {
+      g_signal_handlers_disconnect_by_data (self->impl, self);
+      g_clear_object (&self->impl);
+    }
+
+  self->impl = g_object_ref (new_impl);
+  g_signal_connect (new_impl, "save-hint", G_CALLBACK (save_hint_cb), self);
+  g_signal_connect (new_impl, "quit", G_CALLBACK (quit_cb), self);
+}
+
+static void
+xdp_save_restore_set_property (GObject      *object,
+                               guint         prop_id,
+                               const GValue *value,
+                               GParamSpec   *pspec)
+{
+  XdpSaveRestore *self = XDP_SAVE_RESTORE (object);
+
+  switch (prop_id)
+    {
+    case PROP_IMPL:
+      xdp_save_restore_set_impl (self, g_value_get_object (value));
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+xdp_save_restore_get_property (GObject    *object,
+                               guint       prop_id,
+                               GValue     *value,
+                               GParamSpec *pspec)
+{
+  XdpSaveRestore *self = XDP_SAVE_RESTORE (object);
+
+  switch (prop_id)
+    {
+    case PROP_IMPL:
+      g_value_set_object (value, self->impl);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+xdp_save_restore_class_init (XdpSaveRestoreClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = xdp_save_restore_finalize;
+  object_class->set_property = xdp_save_restore_set_property;
+  object_class->get_property = xdp_save_restore_get_property;
+
+  g_object_class_install_property (object_class, PROP_IMPL,
```

Comment:
FWIW, I don't think there is a good reason to use gobject props here at all. The portals are always final, they never will need to be bindable, etc. Just use a `xdp_save_restore_new` instead.